core/src/main/scala/org/bitcoins/core/protocol/dlc/models/DLCPayoutCurve.scala

```diff
@@ -1,57 +1,45 @@
 package org.bitcoins.core.protocol.dlc.models
 
 import org.bitcoins.core.currency.{CurrencyUnit, Satoshis}
-import org.bitcoins.core.protocol.tlv.{PayoutFunctionV0TLV, TLVPoint}
+import org.bitcoins.core.protocol.tlv._
 import org.bitcoins.core.util.{Indexed, NumberUtil}
 
 import scala.math.BigDecimal.RoundingMode
+import scala.util.{Failure, Success, Try}
 
 /** A DLC payout curve defined by piecewise interpolating points */
-case class DLCPayoutCurve(points: Vector[OutcomePayoutPoint]) {
-  require(points.init.zip(points.tail).forall { case (p1, p2) =>
-            p1.outcome < p2.outcome
-          },
-          s"Points must be ascending: $points")
-
-  def toTLV: PayoutFunctionV0TLV = {
-    PayoutFunctionV0TLV(points.map { point =>
-      TLVPoint(point.outcome,
-               point.roundedPayout,
-               point.extraPrecision,
-               point.isEndpoint)
-    })
-  }
-
-  /** These points (and their indices in this.points) represent the endpoints
-    * between which interpolation happens.
-    * In other words these endpoints define the pieces of the piecewise function.
-    */
-  lazy val endpoints: Vector[Indexed[OutcomePayoutPoint]] =
-    Indexed(points).filter(_.element.isEndpoint)
-
-  /** This Vector contains the function pieces between the endpoints */
-  lazy val functionComponents: Vector[DLCPayoutCurvePiece] = {
-    endpoints.init.zip(endpoints.tail).map { // All pairs of adjacent endpoints
-      case (Indexed(_, index), Indexed(_, nextIndex)) =>
-        DLCPayoutCurvePiece(points.slice(index, nextIndex + 1))
-    }
+case class DLCPayoutCurve(pieces: Vector[DLCPayoutCurvePiece])
+    extends TLVSerializable[PayoutFunctionV0TLV] {
+
+  val endpoints: Vector[OutcomePayoutPoint] = {
+    pieces.map(_.leftEndpoint).:+(pieces.last.rightEndpoint)
+  }
+
+  require(pieces.map(_.rightEndpoint) == endpoints.tail,
+          s"Endpoints must line up: $this")
+
+  //should add invariant here to make sure left endpoint is zero?
```

Comment:
go for it :)

Reply:
i don't think this is right, can't we have signed numbers? I.e. -100 or something like that